internal/buildengine/engine.go

```diff
@@ -377,7 +377,502 @@ func (e *Engine) Modules() []string {
+	dependencies []string
+	isWaiting    bool
+}
+
+func (e *Engine) devBuildLoop(ctx context.Context, period time.Duration) error {
+	logger := log.FromContext(ctx)
+	logger.Debugf("newWatchForModuleChanges")
+
+	// Setup watch channels and initial state
+	watchEvents, cleanup, err := e.setupModuleWatcher(ctx, period)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer cleanup()
+
+	// Track modules waiting to build and their dependencies
+	waitingModules := xsync.NewMapOf[string, moduleBuildState]()
+	readyToBuild := make(chan string, 128) // Using string to pass module names
+
+	// Create a new errgroup with parallelism limit
+	buildGroup, buildCtx := errgroup.WithContext(ctx)
+	buildGroup.SetLimit(e.parallelism)
+
+	// Track which modules have been built successfully
+	builtModules := xsync.NewMapOf[string, *schema.Module]()
+	builtModules.Store("builtin", schema.Builtins())
+
+	// Update schema and set initial module hashes
+	for {
+		select {
+		case event := <-e.deployCoordinator.SchemaUpdates:
+			e.targetSchema.Store(event.schema)
+			continue
+		default:
+		}
+		break
+	}
+
+	moduleHashes := map[string][]byte{}
+	for _, sch := range e.targetSchema.Load().InternalModules() {
+		hash, err := computeModuleHash(sch)
+		if err != nil {
+			return errors.Wrapf(err, "compute hash for %s failed", sch.Name)
+		}
+		moduleHashes[sch.Name] = hash
+	}
+
+	go func() {
+		for {
+			select {
+			case <-buildCtx.Done():
+				return
+			case moduleName := <-readyToBuild:
+				buildGroup.Go(func() error {
+					return errors.WithStack(e.buildAndDeploy(ctx, moduleName, builtModules, waitingModules, readyToBuild))
+				})
+			}
+		}
+	}()
+
+	// Main event loop
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.WithStack(ctx.Err())
+		case event := <-watchEvents:
+			if err := e.handleWatchEvent(ctx, event, waitingModules, readyToBuild); err != nil {
+				logger.Errorf(err, "Error handling watch event")
+			}
+
+		case event := <-e.deployCoordinator.SchemaUpdates:
+			if err := e.handleSchemaUpdate(ctx, event, waitingModules, readyToBuild, moduleHashes); err != nil {
+				logger.Errorf(err, "Error handling schema update")
+			}
+
+		case event := <-e.rebuildEvents:
+			if err := e.handleRebuildEvent(ctx, event, waitingModules, readyToBuild, builtModules); err != nil {
+				logger.Errorf(err, "Error handling rebuild event")
+			}
+		}
+	}
+}
+
+func (e *Engine) buildAndDeploy(ctx context.Context, moduleName string, builtModules *xsync.MapOf[string, *schema.Module], waitingModules *xsync.MapOf[string, moduleBuildState], readyToBuild chan<- string) error {
+	logger := log.FromContext(ctx)
+
+	// Build the module
+	if err := e.buildModule(ctx, moduleName, builtModules); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Store schema after successful build
+	if sch, ok := e.GetModuleSchema(moduleName); ok {
+		builtModules.Store(moduleName, sch)
+	}
+
+	// Deploy the module
+	if err := e.deployModule(ctx, moduleName); err != nil {
```

Comment:
It's more complex but technically we can start building other modules as soon as their deps are built, which would potentially provide a small speedup.

Reply:
yeah true! I'll work on that tomorrow to see if I can eek out some more speed improvements.

I did notice that with this, we  already see more consistent build times and they are a bit faster than the current setup.